File: 프로그래머스/0/181918. 배열 만들기 4/배열 만들기 4.py
# 간단한 풀이
def solution(arr):
    stk = []

    for i in range(len(arr)):
        while stk and stk[-1] >= arr[i]: # 스택이 비어있지 않고, 스택의 맨 위에 있는 값이 현재 숫자 arr[i]보다 크거나 같은 경우에는 반복
            stk.pop()
        stk.append(arr[i])
    return stk

# 내 풀이
def solution(arr):
    stk = []
    i = 0
    
    while i < len(arr): 
        if len(stk) == 0:
            stk.append(arr[i])
            i += 1
        elif len(stk) != 0:
            if stk[-1] < arr[i]:
                stk.append(arr[i])
                i += 1
            elif stk[-1] >= arr[i]:
                stk.pop()
        
    return stk

File: 프로그래머스/0/181918. 배열 만들기 4/배열 만들기 4.java
import java.util.*;

class Solution {
    public int[] solution(int[] arr) {
        ArrayList<Integer> stk = new ArrayList<>();
        
        int i = 0;
        while(i < arr.length){
            if(stk.isEmpty()){
                stk.add(arr[i]);
                i++;
            }
            else if(stk.size() != 0 && stk.get(stk.size() - 1) < arr[i]){
                stk.add(arr[i]);
                i++;
            }
            else if(stk.size() != 0 && stk.get(stk.size() - 1) >= arr[i]){
                stk.remove(stk.size() - 1);
            }
        }
        
        return stk.stream().mapToInt(num -> num).toArray();
    }
}